LoggerExtensions.MessageFormatter takes an exception as argument but ignores it. - TraceSourceProvider should take exception into account, even if the formatter is not null.

### Describe the bug
We are using Asp.net Core (3.1) and the "TraceSourceProvider" (`logging.AddTraceSource(sourceSwitchName)`, which in turn uses the `Microsoft.Extensions.Logging.TraceSource.TraceSourceLogger` --> the TraceSourceLogger relies on the given MessageFormatter and does nothing withe the given exception.
Because `Func<TState, Exception, string> formatter` of the `ILogger` interface is not null the `TraceSourceLogger` "thinks" the exception has already been taken into account by the formatter and does not add the exception (message/stacktrace,...) to the message that will be finally logged. 

We are using Microsoft.Extensions.Logging 3.1.3, should this be fixed in the MessageFormatter method or should the TraceSourceLogger take care to add exception details?

### To Reproduce
Steps to reproduce the behavior:
1. Configure Asp.Net Core logging to use the TraceSourceLogger (`logging.AddTraceSource(sourceSwitchName)`)
2. Use `ILogger<SomeClass>` and log an exception, e.q `logger.LogError("some message", someException);`
3. Look at the log file, the exception is missing, the message ("some message") is available

### Expected behavior
If an exception is available it should be taken care of in the `MessageFormatter  ` and therfore the exception should be visible in the log file.

@ericstj 

Can someone please help me to compile the Microsoft.Extensions.Logging.sln

Here is what I did so far:

build.cmd clr+libs -rc Release (finished with 0 warnings and 0 errors)
build.cmd -vs Microsoft.Extensions.Logging
As soon as VS 2019 (16.8 preview 3) opens I can't compile the solution or run tests:
Error:
1>------ Build started: Project: TestUtilities, Configuration: Debug Any CPU ------
Error occurred while restoring NuGet packages: Invalid restore input. Duplicate frameworks found: 'net5.0, net461, net5.0'. Input files: C:\Dev\Git\runtime\src\libraries\Microsoft.Extensions.Logging\tests\Common\Microsoft.Extensions.Logging.Tests.csproj.
1>CSC : error CS0006: Metadata file 'C:\Dev\Git\runtime\artifacts\bin\System.Runtime.CompilerServices.Unsafe\net45-Debug\System.Runtime.CompilerServices.Unsafe.dll' could not be found
1>Done building project "TestUtilities.csproj" -- FAILED.
1>TestUtilities -> C:\Dev\Git\runtime\artifacts\bin\TestUtilities\net5.0-Debug\TestUtilities.dll
2>------ Build started: Project: Microsoft.Extensions.Logging.Tests, Configuration: Debug Any CPU ------

It seems that VS(nuget restore) are confused with $(NetCoreAppCurrent)-Windows_NT;$(NetCoreAppCurrent).

As there no way make Visual Studio work

--- a/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
@@ -42,7 +42,7 @@ namespace Microsoft.Extensions.Logging.TraceSource
 
             if (exception != null)
             {
-                string exceptionDelimiter = !string.IsNullOrEmpty(message) ? Environment.NewLine : string.Empty;
+                string exceptionDelimiter = string.IsNullOrEmpty(message) ? string.Empty : Environment.NewLine;
                 message += exceptionDelimiter + "Error: " + exception;
             }
 

--- a/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Diagnostics;
+using System.Globalization;
 using DiagnosticsTraceSource = System.Diagnostics.TraceSource;
 
 namespace Microsoft.Extensions.Logging.TraceSource
@@ -35,13 +36,20 @@ namespace Microsoft.Extensions.Logging.TraceSource
                 }
                 if (exception != null)
                 {
-                    message += Environment.NewLine + exception;
+                    message += string.Format(CultureInfo.InvariantCulture, "{0}{1}", Environment.NewLine, exception);
                 }
             }
-            if (!string.IsNullOrEmpty(message))
+
+            if (string.IsNullOrEmpty(message) && exception == null)
+                return;
+
+            if (exception != null)
             {
-                _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
+                string exceptionDelimiter = !string.IsNullOrEmpty(message) ? Environment.NewLine : string.Empty;
+                message += string.Format(CultureInfo.InvariantCulture, "{0}Error: {1}", exceptionDelimiter, exception);
             }
+
+            _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
         }
 
         public bool IsEnabled(LogLevel logLevel)

--- a/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
@@ -3,7 +3,6 @@
 
 using System;
 using System.Diagnostics;
-using System.Globalization;
 using DiagnosticsTraceSource = System.Diagnostics.TraceSource;
 
 namespace Microsoft.Extensions.Logging.TraceSource
@@ -23,7 +22,9 @@ namespace Microsoft.Extensions.Logging.TraceSource
             {
                 return;
             }
+
             string message = string.Empty;
+
             if (formatter != null)
             {
                 message = formatter(state, exception);
@@ -34,10 +35,6 @@ namespace Microsoft.Extensions.Logging.TraceSource
                 {
                     message += state;
                 }
-                if (exception != null)
-                {
-                    message += string.Format(CultureInfo.InvariantCulture, "{0}{1}", Environment.NewLine, exception);
-                }
             }
 
             if (string.IsNullOrEmpty(message) && exception == null)
@@ -46,7 +43,7 @@ namespace Microsoft.Extensions.Logging.TraceSource
             if (exception != null)
             {
                 string exceptionDelimiter = !string.IsNullOrEmpty(message) ? Environment.NewLine : string.Empty;
-                message += string.Format(CultureInfo.InvariantCulture, "{0}Error: {1}", exceptionDelimiter, exception);
+                message += exceptionDelimiter + "Error: " + exception;
             }
 
             _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);

--- a/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.TraceSource/src/TraceSourceLogger.cs
@@ -29,24 +29,21 @@ namespace Microsoft.Extensions.Logging.TraceSource
             {
                 message = formatter(state, exception);
             }
-            else
+            else if (state != null)
             {
-                if (state != null)
-                {
-                    message += state;
-                }
+                message += state;
             }
-
-            if (string.IsNullOrEmpty(message) && exception == null)
-                return;
 
             if (exception != null)
             {
-                string exceptionDelimiter = string.IsNullOrEmpty(message) ? string.Empty : Environment.NewLine;
-                message += exceptionDelimiter + "Error: " + exception;
+                string exceptionDelimiter = string.IsNullOrEmpty(message) ? string.Empty : " " ;
+                message += exceptionDelimiter + exception;
             }
 
-            _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
+            if (!string.IsNullOrEmpty(message))
+            {
+                _traceSource.TraceEvent(GetEventType(logLevel), eventId.Id, message);
+            }
         }
 
         public bool IsEnabled(LogLevel logLevel)

--- a/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
+++ b/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
@@ -1,7 +1,11 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using System.Diagnostics;
+
+using Moq;
+
 using Xunit;
 
 namespace Microsoft.Extensions.Logging.Test
@@ -51,9 +55,36 @@ namespace Microsoft.Extensions.Logging.Test
                 .AddTraceSource(secondSwitch));
 
             var logger = factory.CreateLogger("Test");
-
+            
             // Assert
             Assert.Equal(expected, logger.IsEnabled(LogLevel.Information));
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public static void Log_Shoud_Add_Exception_To_Message_Whether_Formatter_Is_Null_Or_Not(bool shouldFormatterBeNull)
+        {
+            // Arrange
+            Mock<TraceListener> traceListener = new Mock<TraceListener>();
+            SourceSwitch sourceSwitch = new SourceSwitch("TestSwitch") {Level = SourceLevels.All};
+
+            ILoggerFactory factory = TestLoggerBuilder.Create(builder => builder.AddTraceSource(sourceSwitch, traceListener.Object));
+            ILogger logger = factory.CreateLogger("Test");
+
+            const LogLevel logLevel = LogLevel.Information;
+            EventId eventId = new EventId(1);
+            const string message = "some log message";
+            Exception exception = new Exception("Some error occurred");
+            Func<string, Exception, string> formatter = shouldFormatterBeNull ? (Func<string, Exception, string>)null : (value, passedException) => value;
+
+            string expectedMessage = $"{message}{Environment.NewLine}Error: {exception}";
+
+            // Act
+            logger.Log(logLevel, eventId, message, exception, formatter);
+
+            // Assert
+            traceListener.Verify(listener => listener.TraceEvent(It.IsAny<TraceEventCache>(), It.IsAny<string>(), It.IsAny<TraceEventType>(), It.IsAny<int>(), expectedMessage), Times.Once);
         }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
+++ b/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
@@ -3,7 +3,6 @@
 
 using System;
 using System.Diagnostics;
-
 using Moq;
 
 using Xunit;

--- a/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
+++ b/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
@@ -49,7 +49,6 @@ namespace Microsoft.Extensions.Logging.Test
             secondSwitch.Level = second;
 
             // Act
-
             var factory = TestLoggerBuilder.Create(builder => builder
                 .AddTraceSource(firstSwitch)
                 .AddTraceSource(secondSwitch));

--- a/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
+++ b/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
@@ -53,7 +53,7 @@ namespace Microsoft.Extensions.Logging.Test
                 .AddTraceSource(secondSwitch));
 
             var logger = factory.CreateLogger("Test");
-            
+
             // Assert
             Assert.Equal(expected, logger.IsEnabled(LogLevel.Information));
         }

--- a/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
+++ b/src/libraries/Microsoft.Extensions.Logging/tests/Common/TraceSourceLoggerTest.cs
@@ -77,7 +77,7 @@ namespace Microsoft.Extensions.Logging.Test
             Exception exception = new Exception("Some error occurred");
             Func<string, Exception, string> formatter = shouldFormatterBeNull ? (Func<string, Exception, string>)null : (value, passedException) => value;
 
-            string expectedMessage = $"{message}{Environment.NewLine}Error: {exception}";
+            string expectedMessage = $"{message} {exception}";
 
             // Act
             logger.Log(logLevel, eventId, message, exception, formatter);

